group_utils/javascript/automation/lib/javascript/community/groupUtils.js

```diff
@@ -0,0 +1,138 @@
+  this.log = Java.type('org.slf4j.LoggerFactory').getLogger('org.openhab.model.script.Rules.GroupUtils')
+  this.log.debug('Building groupUtils instance.')
+}
+
+/**
+ * Get the direct members of a group.
+ *
+ * @param {*} group Name of the openHAB group
+ * @param {*} characteristic Which characteristic of the members you get, valid: name (default), label, state
+ */
+GroupUtils.prototype.getMembers = function (group, characteristic) {
+  this.log.debug('Getting direct members of group ' + group)
+  var groupMembers = Array()
+  itemRegistry.getItem(group).getMembers().stream().forEach(function (item) {
+    if (characteristic === 'label') {
+      groupMembers.push(item.getLabel())
+    } else if (characteristic === 'state') {
+      groupMembers.push(item.getState())
+    } else {
+      groupMembers.push(item.getName())
+    }
+  })
+  // return the array
+  return groupMembers
+}
+
+/**
+ * Get all (also childs) members of a group, the group items are excluded from this list.
+ *
+ * @param {*} group Name of the openHAB group
+ * @param {*} characteristic Which characteristic of the members you get, valid: name (default), label, state
+ */
+GroupUtils.prototype.getAllMembers = function (group, characteristic) {
+  this.log.debug('Getting all members of group ' + group)
+  var groupAllMembers = Array()
+  itemRegistry.getItem(group).getAllMembers().stream().forEach(function (item) {
+    if (characteristic === 'label') {
+      groupAllMembers.push(item.getLabel())
+    } else if (characteristic === 'state') {
+      groupAllMembers.push(item.getState())
+    } else {
+      groupAllMembers.push(item.getName())
+    }
+  })
+  // return the array
+  return groupAllMembers
+}
+
+/**
+ * Perform arithmetic operations on multiple states.
+ *
+ * @param {*} items Array of states
+ * @param {*} func Arithmetic function to perform, valid: SUM, AVG, MIN, MAX
+ */
+GroupUtils.prototype.arithmetic = function (items, func) {
+  if (func === 'SUM') {
+    var sum = items[0]
+    for (var i = 1; i < items.length; i++) {
+      sum += items[i]
+    }
+    return sum
+  } else if (func === 'AVG') {
+    var sum = items[0]
+    for (var i = 1; i < items.length; i++) {
+      sum += items[i]
+    }
+    var avg = sum / items.length
+    return avg
+  } else if (func === 'MIN') {
+    var min = items[0]
+    for (var i = 1; i < items.length; ++i) {
+      if (items[i] < min) {
+        min = items[i]
+      }
+    }
+    return min
+  } else if (func === 'MAX') {
+    var max = items[0]
+    for (var i = 1; i < items.length; ++i) {
+      if (items[i] > max) {
+        max = items[i]
+      }
+    }
+    return max
+  }
+}
+
+/**
+ * Returns the number of states matching the given comparison expression.
+ *
+ * @param {*} items Array of states
+ * @param {*} op Comparison operator, valid: equal, notEqual, larger, largerEqual, smaller, smallerEqual
+ * @param {*} comp State to compare with
+ */
+GroupUtils.prototype.count = function (items, op, comp) {
```

Comment:
Don't limit what comparisons the users can choose. Have them pass a comparison function of their own devising. Then they can use a comparison as complex as they want including compound comparisons.

It also makes the implementation much simpler.

Also, again if this is a Group utility, pass in the Group name, not an array of Items.

```
// let's be more descriptive as `count` is likely to collide with a variable name the user is already using in their rule
context.countMembers = function(group, compareFunc){
  return context.ir.getItem(group)
                             .members()
                             .stream()
                             .filter(compareFunc)
                             .collect(Collectors.counting());
};
```

To call the function for equals ON the user would use 

```
countMembers("MyGroup", function(i) { return i.state == ON; });
```